src/output.cpp

```diff
@@ -0,0 +1,386 @@
+    if (populations_.find(populationName) == populations_.end()) {
+        populations_.emplace(populationName, Population{filename_, populationName});
+    }
+
+    return populations_.at(populationName);
+}
+
+auto SpikeReader::operator[](const std::string& populationName) const -> const Population& {
+    return getPopulation(populationName);
+}
+
+SpikeReader::Population::Spikes SpikeReader::Population::get(const Selection& node_ids,
+                                                             double tstart,
+                                                             double tstop) const {
+    tstart = tstart < 0 ? tstart_ : tstart;
+    tstop = tstop < 0 ? tstop_ : tstop;
+    if (tstart > tstop_ + EPSILON || tstop < tstart_ - EPSILON || tstop < tstart) {
+        return SpikeReader::Population::Spikes{};
+    }
+
+    auto ret = spikes_;
+    filterTimestamp(ret, tstart, tstop);
+
+    if (!node_ids.empty()) {
+        filterNode(ret, node_ids);
+    }
+
+    return ret;
+}
+
+SpikeReader::Population::Sorting SpikeReader::Population::getSorting() const {
+    return sorting_;
+}
+
+SpikeReader::Population::Population(const std::string& filename,
+                                    const std::string& populationName) {
+    H5::File file(filename, H5::File::ReadOnly);
+    auto pop_path = std::string("/spikes/") + populationName;
+    auto pop = file.getGroup(pop_path);
+
+    std::vector<Spike::first_type> node_ids;
+    pop.getDataSet("node_ids").read(node_ids);
+
+    std::vector<Spike::second_type> timestamps;
+    pop.getDataSet("timestamps").read(timestamps);
+
+    if (node_ids.size() != timestamps.size()) {
+        throw std::runtime_error(
+            "In spikes file, 'node_ids' and 'timestamps' does not have the same size.");
+    }
+
+    std::transform(std::make_move_iterator(node_ids.begin()),
+                   std::make_move_iterator(node_ids.end()),
+                   std::make_move_iterator(timestamps.begin()),
+                   std::back_inserter(spikes_),
+                   [](Spike::first_type&& node_id, Spike::second_type&& timestamp) {
+                       return std::make_pair(std::move(node_id), std::move(timestamp));
+                   });
+
+    if (pop.hasAttribute("sorting")) {
+        pop.getAttribute("sorting").read(sorting_);
+    }
+
+    if (sorting_ == Sorting::by_time) {
+        tstart_ = timestamps.front();
+        tstop_ = timestamps.back();
+    } else {
+        tstart_ = *min_element(timestamps.cbegin(), timestamps.cend());
+        tstop_ = *max_element(timestamps.cbegin(), timestamps.cend());
+    }
+}
+void SpikeReader::Population::filterNode(Spikes& spikes, const Selection& node_ids) const {
+    if (sorting_ == Sorting::by_id) {
+        filterNodeIDSorted(spikes, node_ids);
+    } else {
+        filterNodeIDUnsorted(spikes, node_ids);
+    }
+}
+
+void SpikeReader::Population::filterNodeIDUnsorted(Spikes& spikes,
```

Comment:
Ah, I see what you mean be inplace: these are static, so could just be bare functions.  Perhaps they should be?